File: python/680.py
#!/usr/bin/env python3
"""Project Euler Problem 680: Yarra Gnisrever.

Given an array A of n integers from 0 to n-1, and K operations of the form
"reverse the sub-array from index F_{2n-1} and F_{2n} inclusive (where F_n are
the Fibonacci numbers), find R = sum_{i=0}^{N-1} i A[i] for the final array.

Direct translation from Java reference implementation.
Uses a segment tree with lazy reversal.
"""


def solve():
    N = 10**18
    K = 10**6
    M = 10**9

    def sum_powers_1(n, mod):
        """Sum of 1 + 2 + ... + n mod m."""
        # = n * (n + 1) / 2
        n = n % (2 * mod)  # Ensure no overflow issues
        return (n * (n + 1) // 2) % mod

    def sum_powers_2(n, mod):
        """Sum of 1^2 + 2^2 + ... + n^2 mod m."""
        # = n * (n + 1) * (2n + 1) / 6
        n = n % (6 * mod)  # Ensure divisibility
        return (n * (n + 1) * (2 * n + 1) // 6) % mod

    class Node:
        __slots__ = ['len', 'first', 'diff', 'left', 'right']

        def __init__(self, len_or_left, first_or_right=None, diff=None):
            if first_or_right is None:
                # Node(left, right) constructor - but we pass as positional
                raise ValueError("Use create_leaf or create_internal")
            if diff is not None:
                # Leaf node: Node(len, first, diff)
                self.len = len_or_left
                self.first = first_or_right
                self.diff = diff
                self.left = None
                self.right = None
            else:
                # Internal node: create_internal handles this
                raise ValueError("Use create_internal")

        @staticmethod
        def create_leaf(length, first, diff):
            node = object.__new__(Node)
            node.len = length
            node.first = first
            node.diff = diff
            node.left = None
            node.right = None
            return node

        @staticmethod
        def create_internal(left, right):
            node = object.__new__(Node)
            node.len = left.len + right.len
            node.first = 0  # Not used for internal nodes
            node.diff = 1
            node.left = left
            node.right = right
            return node

        def canonicalize(self):
            if self.left is not None and self.diff == -1:
                left, right = self.left, self.right
                self.left = right.reverse()
                self.right = left.reverse()
                self.diff = 1

        def reverse(self):
            if self.left is None:
                self.first += self.diff * (self.len - 1)
            self.diff *= -1
            return self

        def do_reverse(self, start, end):
            self.canonicalize()
            if self.left is None:
                # Split into [0, start), [start, end) reversed, [end, len)
                self.left = Node.create_leaf(start, self.first, self.diff)
                mid = Node.create_leaf(end - start, self.first + self.diff * (end - 1), -self.diff)
                right_part = Node.create_leaf(self.len - end, self.first + self.diff * end, self.diff)
                self.right = Node.create_internal(mid, right_part)
                self.diff = 1
            elif end <= self.left.len:
                self.left.do_reverse(start, end)
            elif start >= self.left.len:
                self.right.do_reverse(start - self.left.len, end - self.left.len)
            else:
                self.left.ensure_cut_at(start)
                self.right.ensure_cut_at(end - self.left.len)
                left_right, right_left = self.left.right, self.right.left
                self.left.right = right_left.reverse()
                self.right.left = left_right.reverse()
                self.left.len = self.left.left.len + self.left.right.len
                self.right.len = self.right.left.len + self.right.right.len

        def ensure_cut_at(self, index):
            self.canonicalize()
            if self.left is None:
                self.left = Node.create_leaf(index, self.first, self.diff)
                self.right = Node.create_leaf(self.len - index, self.first + self.diff * index, self.diff)
                self.diff = 1
            elif index < self.left.len:
                self.left.ensure_cut_at(index)
                left_left, left_right = self.left.left, self.left.right
                self.left = left_left
                self.right = Node.create_internal(left_right, self.right)
            elif index > self.left.len:
                self.right.ensure_cut_at(index - self.left.len)
                right_left, right_right = self.right.left, self.right.right
                self.right = right_right
                self.left = Node.create_internal(self.left, right_left)

        def R(self, start):
            self.canonicalize()
            if self.len == 0:
                return 0
            if self.left is None:
                # Java: return (start % M * (first % M) % M * (len % M)
                #         + (start * diff + first) % M * sumPowers(len - 1, 1, M)
                #         + diff * sumPowers(len - 1, 2, M)) % M
                #
                # This computes sum_{i=0}^{len-1} (start + i) * (first + diff * i)
                # = sum_{i=0}^{len-1} (start * first + start * diff * i + first * i + diff * i^2)
                # = len * start * first + (start * diff + first) * sum(i) + diff * sum(i^2)
                # where sum(i) = 0+1+...+(len-1) = sumPowers(len-1, 1)
                # and sum(i^2) = 0+1+4+...+(len-1)^2 = sumPowers(len-1, 2)

                term1 = (start % M) * (self.first % M) % M * (self.len % M) % M
                coef = ((start * self.diff + self.first) % M + M) % M
                term2 = coef * sum_powers_1(self.len - 1, M) % M
                term3 = self.diff * sum_powers_2(self.len - 1, M) % M
                return (term1 + term2 + term3) % M
            return (self.left.R(start) + self.right.R(start + self.left.len)) % M

    # Precompute Fibonacci numbers mod N
    F = [0] * (2 * K + 1)
    F[1] = F[2] = 1
    for i in range(3, 2 * K + 1):
        F[i] = (F[i - 2] + F[i - 1]) % N

    # Create initial tree representing [0, 1, 2, ..., N-1]
    tree = Node.create_leaf(N, 0, 1)

    # Perform K reverse operations
    import sys
    sys.setrecursionlimit(5000000)

    for i in range(1, K + 1):
        s = F[2 * i - 1]
        t = F[2 * i]
        if s < t:
            tree.do_reverse(s, t + 1)
        else:
            tree.do_reverse(t, s + 1)

    ans = tree.R(0)
    return ans


def main():
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/680.c
/*
 * Project Euler 680 - Yarra Gnisrever
 *
 * Array [0..N-1], K reverse operations using Fibonacci indices.
 * Compute R = sum i*A[i] mod 10^9.
 *
 * Uses a treap-like implicit tree with lazy reversal and
 * arithmetic-progression leaves.
 * N=10^18, K=10^6, M=10^9.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef long long ll;
typedef __int128 lll;

#define M_VAL 1000000000LL
#define MAX_NODES 20000000

/* Each node represents a range of the array.
 * Leaf: len elements forming an arithmetic progression: first, first+diff, ..., first+diff*(len-1)
 * Internal: left child, right child
 * diff field: for internal, 1 = normal, -1 = reversed
 */
typedef struct {
    ll len;
    ll first;
    int diff;   /* +1 or -1 for both leaf and internal */
    int left;   /* child index, -1 for leaf */
    int right;  /* child index, -1 for leaf */
} Node;

static Node nodes[MAX_NODES];
static int num_nodes = 0;

static int new_leaf(ll len, ll first, int diff) {
    int id = num_nodes++;
    nodes[id].len = len;
    nodes[id].first = first;
    nodes[id].diff = diff;
    nodes[id].left = -1;
    nodes[id].right = -1;
    return id;
}

static int new_internal(int left, int right) {
    int id = num_nodes++;
    nodes[id].len = nodes[left].len + nodes[right].len;
    nodes[id].first = 0;
    nodes[id].diff = 1;
    nodes[id].left = left;
    nodes[id].right = right;
    return id;
}

static void node_reverse(int id) {
    if (nodes[id].left == -1) {
        /* Leaf: first -> first + diff*(len-1), diff -> -diff */
        nodes[id].first += (ll)nodes[id].diff * (nodes[id].len - 1);
    }
    nodes[id].diff *= -1;
}

static void canonicalize(int id) {
    if (nodes[id].left != -1 && nodes[id].diff == -1) {
        int left = nodes[id].left;
        int right = nodes[id].right;
        /* Swap and reverse children */
        node_reverse(right);
        node_reverse(left);
        nodes[id].left = right;
        nodes[id].right = left;
        nodes[id].diff = 1;
    }
}

static void ensure_cut_at(int id, ll index);
static void do_reverse(int id, ll start, ll end);

static void ensure_cut_at(int id, ll index) {
    canonicalize(id);
    if (nodes[id].left == -1) {
        /* Split leaf into [0, index) and [index, len) */
        int l = new_leaf(index, nodes[id].first, nodes[id].diff);
        int r = new_leaf(nodes[id].len - index,
                         nodes[id].first + (ll)nodes[id].diff * index,
                         nodes[id].diff);
        nodes[id].left = l;
        nodes[id].right = r;
        nodes[id].diff = 1;
    } else if (index < nodes[nodes[id].left].len) {
        ensure_cut_at(nodes[id].left, index);
        int ll_node = nodes[nodes[id].left].left;
        int lr_node = nodes[nodes[id].left].right;
        int old_right = nodes[id].right;
        nodes[id].left = ll_node;
        nodes[id].right = new_internal(lr_node, old_right);
    } else if (index > nodes[nodes[id].left].len) {
        ensure_cut_at(nodes[id].right, index - nodes[nodes[id].left].len);
        int rl_node = nodes[nodes[id].right].left;
        int rr_node = nodes[nodes[id].right].right;
        int old_left = nodes[id].left;
        nodes[id].right = rr_node;
        nodes[id].left = new_internal(old_left, rl_node);
    }
    /* else index == left.len, already cut */
}

static void do_reverse(int id, ll start, ll end) {
    canonicalize(id);
    if (nodes[id].left == -1) {
        /* Split leaf into [0, start), [start, end) reversed, [end, len) */
        int left_part = new_leaf(start, nodes[id].first, nodes[id].diff);
        int mid_part = new_leaf(end - start,
                                nodes[id].first + (ll)nodes[id].diff * (end - 1),
                                -nodes[id].diff);
        int right_part = new_leaf(nodes[id].len - end,
                                  nodes[id].first + (ll)nodes[id].diff * end,
                                  nodes[id].diff);
        int right_combined = new_internal(mid_part, right_part);
        nodes[id].left = left_part;
        nodes[id].right = right_combined;
        nodes[id].diff = 1;
    } else {
        ll left_len = nodes[nodes[id].left].len;
        if (end <= left_len) {
            do_reverse(nodes[id].left, start, end);
        } else if (start >= left_len) {
            do_reverse(nodes[id].right, start - left_len, end - left_len);
        } else {
            /* Spans both children */
            ensure_cut_at(nodes[id].left, start);
            ensure_cut_at(nodes[id].right, end - nodes[nodes[id].left].len);

            int left_right = nodes[nodes[id].left].right;
            int right_left = nodes[nodes[id].right].left;

            node_reverse(left_right);
            node_reverse(right_left);

            nodes[nodes[id].left].right = right_left;
            nodes[nodes[id].right].left = left_right;

            nodes[nodes[id].left].len = nodes[nodes[nodes[id].left].left].len +
                                        nodes[nodes[nodes[id].left].right].len;
            nodes[nodes[id].right].len = nodes[nodes[nodes[id].right].left].len +
                                         nodes[nodes[nodes[id].right].right].len;
        }
    }
}

static ll sum_powers_1(ll n) {
    /* Sum of 1 + 2 + ... + n mod M */
    n = n % (2 * M_VAL);
    return (n * (n + 1) / 2) % M_VAL;
}

static ll sum_powers_2(ll n) {
    /* Sum of 1^2 + 2^2 + ... + n^2 mod M */
    n = n % (6 * M_VAL);
    return (n * (n + 1) % (6 * M_VAL) * (2 * n + 1) / 6) % M_VAL;
}

static ll compute_R(int id, ll start) {
    canonicalize(id);
    if (nodes[id].len == 0) return 0;
    if (nodes[id].left == -1) {
        /* Leaf: sum_{i=0}^{len-1} (start+i) * (first + diff*i)
         * = len*start*first + (start*diff + first)*sum(i) + diff*sum(i^2)
         */
        ll len = nodes[id].len;
        ll first = nodes[id].first;
        int diff = nodes[id].diff;

        ll term1 = (start % M_VAL) * (first % M_VAL) % M_VAL * (len % M_VAL) % M_VAL;
        term1 = ((term1 % M_VAL) + M_VAL) % M_VAL;

        ll coef = ((start * diff + first) % M_VAL + M_VAL) % M_VAL;
        ll term2 = coef * sum_powers_1(len - 1) % M_VAL;

        ll term3 = (ll)diff * sum_powers_2(len - 1) % M_VAL;
        term3 = (term3 + M_VAL) % M_VAL;

        return (term1 + term2 + term3) % M_VAL;
    }
    ll left_r = compute_R(nodes[id].left, start);
    ll right_r = compute_R(nodes[id].right, start + nodes[nodes[id].left].len);
    return (left_r + right_r) % M_VAL;
}

int main() {
    ll N = 1000000000000000000LL; /* 10^18 */
    int K = 1000000; /* 10^6 */

    /* Precompute Fibonacci numbers mod N */
    ll *F = (ll *)calloc(2 * K + 2, sizeof(ll));
    F[1] = 1; F[2] = 1;
    for (int i = 3; i <= 2 * K; i++) {
        F[i] = (F[i - 2] + F[i - 1]) % N;
    }

    /* Create initial tree */
    int root = new_leaf(N, 0, 1);

    for (int i = 1; i <= K; i++) {
        ll s = F[2 * i - 1];
        ll t = F[2 * i];
        if (s < t) {
            do_reverse(root, s, t + 1);
        } else {
            do_reverse(root, t, s + 1);
        }
    }

    ll ans = compute_R(root, 0);
    printf("%lld\n", ans);

    free(F);
    return 0;
}
